src/sql/src/plan/statement/show.rs

```diff
@@ -924,29 +924,31 @@ fn humanize_sql_for_show_create(
         // source is altered.
         Statement::CreateSource(stmt) => {
             // Collect all current subsource references.
-            //
-            // TODO(#24843): this structure will need to change because we will
-            // have multiple references to the same upstream table.
-            let mut curr_references: BTreeMap<_, _> = catalog
-                .get_item(&id)
-                .used_by()
-                .into_iter()
-                .filter_map(|subsource| {
-                    let item = catalog.get_item(subsource);
-                    item.subsource_details().map(|(_id, reference)| {
-                        let name = item.name();
-                        (
-                            reference.clone(),
-                            ResolvedItemName::Item {
-                                id: item.id(),
-                                qualifiers: name.qualifiers.clone(),
-                                full_name: catalog.resolve_full_name(name),
-                                print_id: false,
-                            },
-                        )
+            let mut curr_references = BTreeMap::new();
+            for dependent in catalog.get_item(&id).used_by().into_iter() {
+                let item = catalog.get_item(dependent);
+                let reference = match item.subsource_details() {
+                    // Any item that uses this source and has subsource details
+                    // should be one of its subsources; however, if it's not,
+                    // this is low-stakes to just skip past it.
+                    Some((primary_source_id, reference)) if primary_source_id == id => reference,
+                    // Not a subsource of this source.
+                    _ => continue,
+                };
+
+                curr_references
+                    .entry(reference.clone())
+                    .and_modify(|(id, name)| {
+                        // Keep the entry with the maximum ID; from a
+                        // perspective of subsource demuxing this must be the
+                        // "widest" and latest version of the subsource.
+                        if *id < *dependent {
```

Comment:
this (that the max ID corresponds to the widest subsource) seems overly brittle / restrictive, and I'm not sure aligns well with a potential schema change worklow that a customer may go through:

consider the case where a customer has a table with 4 columns: A, B, C, D that is currently being ingested as subsource foo.

Then they make an application change where they decide to introduce column E, a slightly better representation of the data they currently put into column B. To support a gradual migration of their application code they add column E, backfill it and have their application populate both columns. After they finish migrating all their application code they then drop column B.

From a materialize perspective, when they add column E they'd ideally like to define a new subsource export that contains A, C, D, E since they know B will be deprecated soon, and they can start migrating their downstream mat-view/queries to use this new export in the meantime. They don't want to have to create a 3rd subsource export once column B is eventually dropped.

Reply:
This is in `SHOW CREATE SOURCE`, which shows you how to create a source that will return you to the state that you are currently in to the best of our ability. To the best of our knowledge, the subsource that most accurately reflects the upstream schema is the last one we saw. There's nothing brittle or restrictive about this line of code––from a purification perspective, this subsource most accurately represents the upstream table's schema as far as we know.

What you've described is a feature out of scope of what this PR proposes we support (nor is it a feature that customers have requested AFAIK). But nothing in this PR's design precludes adding a more complex version of this feature (e.g. what you've described or allowing more complex projections).

However, the usecase for supporting dropped columns is a feature very different from this one––it would require some very different logic and, if it were supported, this code would still be accurate.

Imagine we devise a scheme that allows multiple projections over a dataflow's output using `projection: Vec<usize>`. To support dropping columns, we would need something like `Vec<Option<usize>>`, to express that this projection might simply not be able to find its column in the `output_index`. So then, in the case that a column cannot be found (i.e. the project is `None`) what do we do then? Produce `NULL` values in its place? That is very worrying to introduce data on the user's behalf. Imagine there are e.g. materialized views that depend on this source, and that in the "native"/expected schema `NULL` has a special meaning, e.g. there is a JOIN `NULL` values that means that the customer's loan is denied, or their home has been foreclosed on. Suddenly, because of this schema change, all new rows make it look like the 2008 financial crisis. I don't think that's tenable.

Ok––so, if you have any columns where the value is `None`, you throw a definite error? This is already what we do in the face of incompatible schema changes, and this definite error cascades down to all of its dependents, rendering them unqueryable. Asking a user to drop unqueryable objects doesn't seem that unreasonable IMO.

Imagine they don't want to drop the objects for whatever reason (they can handle definite errors, but not missing object errors), so we want to provide users the option to cut over between subsource schemas (we've already eliminated `projection` as a viable alternative). This feature would not look like demuxing an output index into multiple persist shards. We would have to forcibly stop outputting values to the source's persist shard (maybe a new `bool` flag on `SourceExport` called `do_not_output`), and we could let users instead use some other schema. That feature would need to be distinct from this in some way because we wouldn't want a user typing `ADD SUBSOURCE`, being slightly mistaken about their schemas, and rendering a bunch of their current objects unqueryable. I think this could be a useful feature (`ALTER SOURCE...ADD SUBSOURCE foo WITH (STOP RENDERING CURRENT REFERENCES)`), but its semantics are different from letting users do simple blue/green to ingest new columns.

I'll update the description of this PR to clearly express that it is meant only to let users expand their subsource's schemas to ingest new columns and isn't meant to do anything else.

Reply:
That all makes sense -- thanks for explaining your reasoning and updating the PR description with this context.

> What you've described is a feature out of scope of what this PR proposes we support

It would have been helpful if the PR had a description of what its scope/goals were, and how it was different from the linked 'design' comment, before being put out for review since it's hard to know how you arrived at the current implementation otherwise. There also should likely be some product input for changes like this since it affects the limitations that the associated new feature will have.

Reply:
Since last year, this feature has been discussed many times and for multiple hours. I simply forgot that things had drifted. A good reminder that not everyone has all of the context. Maybe a good approach if you see issues like this in the future from me would be to ask for more clarity on the context of the change, or request a quick five-minute sync because things look different than you expect.